File: src/main/java/my/maryvk/maryvkweb/service/VkServiceImpl.java
package my.maryvk.maryvkweb.service;

import com.vk.api.sdk.client.VkApiClient;
import com.vk.api.sdk.client.actors.UserActor;
import com.vk.api.sdk.exceptions.ApiException;
import com.vk.api.sdk.exceptions.ClientException;
import com.vk.api.sdk.objects.users.UserXtrCounters;
import lombok.extern.java.Log;
import my.maryvk.maryvkweb.domain.RelationType;
import my.maryvk.maryvkweb.domain.User;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Service("vk")
@Log
public class VkServiceImpl implements VkService {

    private final VkApiClient vk;
    private final UserActor owner;

    private final UserService userService;

    public VkServiceImpl(VkApiClient vk, UserActor owner, UserService userService) {
        this.vk = vk;
        this.owner = owner;
        this.userService = userService;
    }

    @Override
    public List<Integer> getConnectedIds(Integer userId, RelationType relationType) {
        List<Integer> ids = relationType == RelationType.FRIEND ? getFriendIds(userId) : getFollowerIds(userId);
        if (ids != null) {
            if (!tryUpdateUsers(ids))
                return null;
        }
        return ids;
    }

    private List<Integer> getFriendIds(Integer userId) {
        try {
            return vk.friends().get(owner).userId(userId).execute().getItems();
        } catch (ApiException | ClientException e) {
            log.warning("Unable to get friends: " + e.getMessage());
            return null;
        }
    }

    private List<Integer> getFollowerIds(Integer userId) {
        try {
            return vk.users().getFollowers(owner).userId(userId).execute().getItems();
        } catch (ApiException | ClientException e) {
            log.warning("Unable to get followers: " + e.getMessage());
            return null;
        }
    }

    private boolean tryUpdateUsers(List<Integer> ids) {
        List<Integer> toUpdate = ids.stream()
                .filter(id -> !userService.exists(id))
                .collect(Collectors.toList());
        if (!toUpdate.isEmpty()) {
            List<User> usersFromVk = getUsersFromVk(toUpdate);
            if (usersFromVk == null)
                return false;
            userService.save(usersFromVk);
        }
        return true;
    }

    private List<User> getUsersFromVk(List<Integer> ids) {
        try {
            List<String> idsStr = ids.stream().map(Object::toString).collect(Collectors.toList());
            List<UserXtrCounters> vkUsers = vk.users().get(owner).userIds(idsStr).execute();
            return vkUsers.stream().map(this::createUser).collect(Collectors.toList());
        } catch (ApiException | ClientException e) {
            log.warning("Unable to get users: " + e.getMessage());
            return null;
        }
    }

    private User createUser(UserXtrCounters u) {
        return User.builder()
                .id(u.getId())
                .firstName(u.getFirstName())
                .lastName(u.getLastName())
                .build();
    }

    @Override
    public User getUser(Integer id) {
        User user = userService.findOne(id);
        if (user == null) {
            tryUpdateUsers(Collections.singletonList(id));
            user = userService.findOne(id);
        }
        return user;
    }
}

File: src/main/java/my/maryvk/maryvkweb/service/VkServiceImpl.kt
package my.maryvk.maryvkweb.service

import com.vk.api.sdk.client.VkApiClient
import com.vk.api.sdk.client.actors.UserActor
import com.vk.api.sdk.exceptions.ApiException
import com.vk.api.sdk.exceptions.ClientException
import com.vk.api.sdk.objects.users.UserXtrCounters
import my.maryvk.maryvkweb.domain.RelationType
import my.maryvk.maryvkweb.domain.User
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.lang.System.currentTimeMillis
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.locks.ReentrantLock

@Service("vk") class VkServiceImpl(
        private val vk: VkApiClient,
        private val owner: UserActor,
        private val userService: UserService,
        @Value("\${vk.api-call-delay}")
        private val apiCallRange: Int
) : VkService {

    override fun getConnectedIds(userId: Int, relationType: RelationType): List<Int>? {
        val ids = if (relationType === RelationType.FRIEND) getFriendIds(userId) else getFollowerIds(userId)
        if (ids != null)
            if (!tryUpdateUsers(ids))
                return null
        return ids
    }

    private fun getFriendIds(userId: Int?): List<Int>? {
        try {
            return executeVkApiCall { vk.friends().get(owner).userId(userId!!).execute().items }
        } catch (e: ApiException) {
//            log.warning("Unable to execute friends: " + e.message)
            return null
        } catch (e: ClientException) {
//            log.warning("Unable to execute friends: " + e.message)
            return null
        }

    }

    private fun getFollowerIds(userId: Int?): List<Int>? {
        try {
            return executeVkApiCall { vk.users().getFollowers(owner).userId(userId!!).execute().items }
        } catch (e: ApiException) {
//            log.warning("Unable to execute followers: " + e.message)
            return null
        } catch (e: ClientException) {
//            log.warning("Unable to execute followers: " + e.message)
            return null
        }

    }

    private fun tryUpdateUsers(ids: List<Int>): Boolean {
        val toUpdate = ids
                .filter { id -> !userService.exists(id) }
                .toList()
        if (!toUpdate.isEmpty()) {
            val usersFromVk = getUsersFromVk(toUpdate) ?: return false
            userService.save(usersFromVk)
        }
        return true
    }

    private fun getUsersFromVk(ids: List<Int>): List<User>? {
        try {
            val idsStr = ids.map { it.toString() }.toList()
            val vkUsers = executeVkApiCall { vk.users().get(owner).userIds(idsStr).execute() }
            return vkUsers.map { createUser(it) }.toList()
        } catch (e: ApiException) {
//            log.warning("Unable to execute users: " + e.message)
            return null
        } catch (e: ClientException) {
//            log.warning("Unable to execute users: " + e.message)
            return null
        }

    }

    private fun createUser(u: UserXtrCounters): User {
        return User(id = u.id, firstName = u.firstName, lastName = u.lastName)
    }

    private val lastTimeApiUsed = AtomicLong(0)
    private val lock = ReentrantLock(true)
    @Throws(ApiException::class, ClientException::class)
    private fun <T> executeVkApiCall(func: () -> T): T {
        lock.lock()
        while (currentTimeMillis() - lastTimeApiUsed.get() < apiCallRange);
        try {
            return func()
        } finally {
            lastTimeApiUsed.set(currentTimeMillis())
            lock.unlock()
        }
    }

    override fun getUser(id: Int): User {
        var user: User? = userService.findOne(id)
        if (user == null) {
            tryUpdateUsers(listOf(id))
            user = userService.findOne(id)
        }
        return user
    }
}